vulyk/blueprints/gamification/core/queries.py

```diff
@@ -0,0 +1,158 @@
+# -*- coding: utf-8 -*-
+"""
+The factory of achievements. Classes below allow us to query data source we use
+as a source of truth.
+"""
+from bson import ObjectId
+from datetime import date, timedelta
+from collections import namedtuple
+
+from mongoengine.queryset.base import BaseQuerySet
+
+from .rules import Rule, ProjectRule
+
+__all__ = [
+    'RULES',
+    'RuleQueryBuilder',
+    'MongoRuleQueryBuilder',
+    'MongoRuleExecutor'
+]
+
+
+class RuleQueryBuilder:
+    """
+    Abstract query builder. Takes an instance of Rule-type and converts its
+    properties into fully fledged queries to a data source.
+    """
+
+    def __init__(self, rule: Rule) -> None:
+        """
+        :param rule: Actual rule to be calculated.
+        :type rule: Rule
+        """
+
+        self._filter_first = {}
+        self._projection = {}
+        self._filter_second = {}
+        self._group = {}
+
+    def build_for(self, user_id: ObjectId) -> list:
+        """
+        Prepares a pipeline of actions to be passed to data source.
+
+        :param user_id: Current user ID
+        :type user_id: bson.ObjectId
+
+        :returns: List of actions to be done.
+        :rtype: list[dict]
+        """
+        raise NotImplementedError()
+
+
+class MongoRuleQueryBuilder(RuleQueryBuilder):
+    """
+    Implementation of RuleQueryBuilder, bound to MongoDB.
+    """
+    _Pair = namedtuple('Pair', ['key', 'clause'])
+
+    def __init__(self, rule: Rule) -> None:
```

Comment:
W/o comments it is less than 40 LOC and, furthermore, it does a single atomic task.